graylog2-server/src/main/java/org/graylog2/indexer/MongoIndexSetRegistry.java

```diff
@@ -16,32 +16,79 @@
  */
 package org.graylog2.indexer;
 
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
 import org.graylog2.indexer.indexset.IndexSetConfig;
 import org.graylog2.indexer.indexset.IndexSetService;
+import org.graylog2.indexer.indexset.events.IndexSetCreatedEvent;
+import org.graylog2.indexer.indexset.events.IndexSetDeletedEvent;
 import org.graylog2.indexer.indices.TooManyAliasesException;
 
 import javax.inject.Inject;
+import javax.inject.Singleton;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static java.util.Objects.requireNonNull;
 
+@Singleton
 public class MongoIndexSetRegistry implements IndexSetRegistry {
     private final IndexSetService indexSetService;
     private final MongoIndexSet.Factory mongoIndexSetFactory;
 
+    static class IndexSetsCache {
+        private final IndexSetService indexSetService;
+        private AtomicReference<Supplier<List<IndexSetConfig>>> indexSetConfigs;
```

Comment:
Using the supplier guarantees it is refreshed lazily, to prevent every node hammering the collection after a change event.